File: apps/macos/PasteSheets/Domain/Repositories/DirectoryRepository.swift
import Foundation

protocol DirectoryRepository {
    func getAllDirectories() throws -> [DirectoryInfo]
    func createDirectory(name: String) throws -> Int64
    func renameDirectory(oldName: String, newName: String) throws
    func deleteDirectory(name: String) throws
}

final class DirectoryRepositoryImpl: DirectoryRepository {
    private let dataSource: DirectoryDataSource

    init(dataSource: DirectoryDataSource = DirectoryDataSourceImpl()) {
        self.dataSource = dataSource
    }

    func getAllDirectories() throws -> [DirectoryInfo] {
        try dataSource.fetchAll().map(DirectoryInfo.init)
    }

    func createDirectory(name: String) throws -> Int64 {
        try dataSource.insert(name: name)
    }

    func renameDirectory(oldName: String, newName: String) throws {
        try dataSource.rename(oldName: oldName, newName: newName)
    }

    func deleteDirectory(name: String) throws {
        try dataSource.delete(name: name)
    }
}

File: apps/windows/PasteSheet/Domain/Repositories/DirectoryRepository.cs
using PasteSheet.Data.DataSources;
using PasteSheet.Domain.Entities;

namespace PasteSheet.Domain.Repositories;

public interface IDirectoryRepository
{
    List<DirectoryInfo> GetAllDirectories();
    long CreateDirectory(string name);
    void RenameDirectory(string oldName, string newName);
    void DeleteDirectory(string name);
}

public sealed class DirectoryRepository : IDirectoryRepository
{
    private readonly IDirectoryDataSource _dataSource;

    public DirectoryRepository(IDirectoryDataSource dataSource) => _dataSource = dataSource;

    public List<DirectoryInfo> GetAllDirectories() =>
        _dataSource.FetchAll().Select(dto => new DirectoryInfo(dto)).ToList();

    public long CreateDirectory(string name) => _dataSource.Insert(name);

    public void RenameDirectory(string oldName, string newName) =>
        _dataSource.Rename(oldName, newName);

    public void DeleteDirectory(string name) => _dataSource.Delete(name);
}
